DF/data4es: allow skipping stages

mgolosova opened the pull request:
Main changes:
1. `--ignore <stageID>[,<stageID>]` option to skip specific stages.
1. `--list` option to show list of stage IDs (to be used with `--ignore`).
1. `--help` option to show usage info.
1. Minor improvements (e.g. show usage info in case of unknown option).
1. Lots of refactoring:
    * wrap pieces of code into functions;
    * reorder some actions;
    * get rid of hard-coded pid file name.
1. Readability improvements (comments, etc.).

---

The goal is to allow `data4es` to utilize ES `update` functionality when load data.
It will allow to "disable" some ETL process nodes for faster processing and not to lose information written to ES earlier. Of course, it will make high load on the ES server, so must be used with caution; still in some situations it looks like a better solution (not to load external sources; not to request ES for known information if external source is not available or doesn\`t have requested information).

---

ToDo:
 - [x] make Stage 19 to recognize and properly handle "update" records (see #254);
 - [x] add option `--update` to Stage 19 (changing default action from `index` to `update`) (see #254);
- [ ] ~(make stages that might stumble over external sources mark output messages "for update":)~
   - [ ] ~(025_chicagoES (see #261);)~
   - [ ] ~(091_datasetsRucio (see #262);)~
   - [ ] ~(095_datasetInfoAMI;)~
 - [x] allow to disable specific stages in `data4es-start`:
     - [x] add `--skip` option to stages (see #260):
         - [x] 025_chicagoES (see #261);
         - [x] 091_datasetsRucio (see #262);
         - [x] 095_datasetInfoAMI (see #263);
     - [x] add `--ignore STAGE...` option to `data4es-start` (currently in this PR);
 - [x] track #251 and pull its changes here as soon as it is approved/merged to `master`.

mgolosova commented:
Instead of rebasing this branch on the current version of `master`, rewrote its history to remove "dirt" commits (appeared due to later changes in branches, that were previously merged here). Not a super-beautiful decision -- the price to pay for rebase-updates made in other branches (and for the desire to keep the merge history).

mgolosova commented:
Resolve **[WIP]** status for what\`s left to do is not really belong here -- it is up to every individual stage.

Evildoor commented on Utils/Dataflow/run/data4es-start:
```diff
@@ -4,10 +4,39 @@ SCRIPT_NAME=data4es
+  -l, --list  list known stage IDs and exit
+
+  --debug     do not actualy load integrated metadata to the DKB Elasticsearch
+              (will creates files with data flow snapshots, taken between
```
1. "Will create files".
2. Maybe "dataflow", in a single word?

Evildoor replied:
Point 1 was fixed, what about 2? Was it missed, or judged fine as-is?

mgolosova replied:
@Evildoor,
Sorry, I definitely remember writing a reply to your comment, but seems like it never was posted.

> "Will create files".

Fixed in 69d0fa0.

> Maybe "dataflow", in a single word?

We often use "dataflow" to refer to the integration process -- data processing stages, being executed in a proper order. Here I meant the flow of data through the stages, not the stages themselves -- so the two-words spelling was intentional. If it looks confusing, I can replace it with "flow of data".

Evildoor replied:
Aha, I see. Let's leave it as-is then.

mgolosova commented:
@Evildoor,
> 9f5b26b: this commit's description talks about changing "OC" to "09", but also transforms (unrelated to OC/09) line 208 into three lines (due to length exceeding 80 symbols, I presume).

Thank you, fixed now.

> 0e4e8c9 and fb6c387: options `--help` and `--list` are implied, both by general logic and description, to merely output some info and exit. However, some work is actually done and, for example, checking of 095's credentials may fail and lead to script termination without info being displayed.

Good point, thank you; it also appeared that they failed to be properly handled in case of already running process. Fixed now.

Evildoor commented:
> > [0e4e8c9](https://github.com/PanDAWMS/dkb/commit/0e4e8c928e89f9f75b8fcb6711d5192e95f7d0b4) and [fb6c387](https://github.com/PanDAWMS/dkb/commit/fb6c3874d49162abac9afb734ae950cdb00b304d): options `--help` and `--list` are implied, both by general logic and description, to merely output some info and exit. However, some work is actually done and, for example, checking of 095's credentials may fail and lead to script termination without info being displayed.
> 
> Good point, thank you; it also appeared that they failed to be properly handled in case of already running process. Fixed now.

Branches are still created (Line 195) when these options are used, which may lead to similar problems:

```
[vaulov@aiatlas171 run]$ ./data4es-start --help
mkfifo: cannot create fifo Б─≤/afs/cern.ch/user/v/vaulov/Work/dkb/Utils/Dataflow/run/.data4es/.pipe/b_91Б─≥: Operation not permitted
(ERROR) 11-11-2019 15:05:05 (data4es) branch(): Failed to create named pipe for branch b_91
```

mgolosova commented:
@Evildoor,
> > > [0e4e8c9](https://github.com/PanDAWMS/dkb/commit/0e4e8c928e89f9f75b8fcb6711d5192e95f7d0b4) and [fb6c387](https://github.com/PanDAWMS/dkb/commit/fb6c3874d49162abac9afb734ae950cdb00b304d): options `--help` and `--list` are implied, both by general logic and description, to merely output some info and exit. However, some work is actually done and, for example, checking of 095's credentials may fail and lead to script termination without info being displayed.
> > 
> > 
> > Good point, thank you; it also appeared that they failed to be properly handled in case of already running process. Fixed now.
> 
> Branches are still created (Line 195) when these options are used, which may lead to similar problems:
> 
> ```
> [vaulov@aiatlas171 run]$ ./data4es-start --help
> mkfifo: cannot create fifo Б─≤/afs/cern.ch/user/v/vaulov/Work/dkb/Utils/Dataflow/run/.data4es/.pipe/b_91Б─≥: Operation not permitted
> (ERROR) 11-11-2019 15:05:05 (data4es) branch(): Failed to create named pipe for branch b_91
> ```

Should work properly now, please take a look.

Evildoor commented:
> @Evildoor,
> 
> > > > [0e4e8c9](https://github.com/PanDAWMS/dkb/commit/0e4e8c928e89f9f75b8fcb6711d5192e95f7d0b4) and [fb6c387](https://github.com/PanDAWMS/dkb/commit/fb6c3874d49162abac9afb734ae950cdb00b304d): options `--help` and `--list` are implied, both by general logic and description, to merely output some info and exit. However, some work is actually done and, for example, checking of 095's credentials may fail and lead to script termination without info being displayed.
> > > 
> > > 
> > > Good point, thank you; it also appeared that they failed to be properly handled in case of already running process. Fixed now.
> > 
> > 
> > Branches are still created (Line 195) when these options are used, which may lead to similar problems:
> > ```
> > [vaulov@aiatlas171 run]$ ./data4es-start --help
> > mkfifo: cannot create fifo Б─≤/afs/cern.ch/user/v/vaulov/Work/dkb/Utils/Dataflow/run/.data4es/.pipe/b_91Б─≥: Operation not permitted
> > (ERROR) 11-11-2019 15:05:05 (data4es) branch(): Failed to create named pipe for branch b_91
> > ```
> 
> Should work properly now, please take a look.

Yes, works properly now, thank you.

Evildoor commented:
Checked again: while `--help` works correctly, `--list` still does some work and produces errors:

```
[vaulov@aiatlas171 run]$ ./data4es-start -l
./data4es-start: line 116: source: filename argument required
source: usage: source filename [arguments]
(ERROR) 23-01-2020 13:30:22 (data4es) Stage 095: wrong authentication parameters (failed to read certificate files:  and )
09
16
17
19
25
69
91_in
91_out
93
95
```

Obvious difference between `--help` and `--list` is definition of stages used in the latter to get the stage numbers from variable names, which seems to be the source of the problem (aforementioned line 116).

Evildoor commented on Utils/Dataflow/run/data4es-start:
```diff
@@ -119,10 +169,9 @@ mediator() {
-# Create named pipes for dataflow branching
-# ---
+# Define dataflow branch (subchain)
+# Creates entry point for the branch (named pipe) and returns its name
```
Maybe "Create", to make this line into "Do X" form like most other comments?

mgolosova replied:
Done: 56dfca5.

Evildoor replied:
Uh, actually there is also "returns" in the end of the line which should be "return". Sorry, did not see and point it out before.

mgolosova replied:
Thank you, fixed now: f02f272.

Evildoor commented:
Comment to the PR text itself: it seems to me that four boxes starting with "make stages that might ..." should be ticked (and the last one should also link to #263, because #260, #261, #262 and #263 responsible for them were merged successfully. I'm writing this instead of fixing the text myself or ignoring it, in case that I'm wrong and something is actually not finished there.

mgolosova commented:
@Evildoor,
> Checked again: while `--help` works correctly, `--list` still does some work and produces errors:
> ```
> [vaulov@aiatlas171 run]$ ./data4es-start -l
> ./data4es-start: line 116: source: filename argument required
> source: usage: source filename [arguments]
> (ERROR) 23-01-2020 13:30:22 (data4es) Stage 095: wrong authentication parameters (failed to > read certificate files:  and )
> <...>
> ```
> Similar problem arises if stage 095 is added to ignore list.
> Also, the error message part about "certificate files: and " can be puzzling - maybe add a specific error message for the case when files are not defined?

I suggest to fix it in a probably unexpected way: remove handling of the files required for Stage 095 from `data4es-start` (check #314). This way no problem will be detected unless the stage is started -- and as a stage started, it knows what to do with these parameters in both modes ("skip" and normal processing).

Evildoor commented:
> > Checked again: while `--help` works correctly, `--list` still does some work and produces errors:
> > ```
> > [vaulov@aiatlas171 run]$ ./data4es-start -l
> > ./data4es-start: line 116: source: filename argument required
> > source: usage: source filename [arguments]
> > (ERROR) 23-01-2020 13:30:22 (data4es) Stage 095: wrong authentication parameters (failed to > read certificate files:  and )
> > <...>
> > ```
> > 
> > 
> > Similar problem arises if stage 095 is added to ignore list.
> > Also, the error message part about "certificate files: and " can be puzzling - maybe add a specific error message for the case when files are not defined?
> 
> I suggest to fix it in a probably unexpected way: remove handling of the files required for Stage 095 from `data4es-start` (check #314). This way no problem will be detected unless the stage is started -- and as a stage started, it knows what to do with these parameters in both modes ("skip" and normal processing).

Looks good to me.

mgolosova commented:
@Evildoor,
> > I suggest to fix it <...> (check #314). 
>
> Looks good to me.

Alright, #314 is merged now and this PR is updated.
Looks like all the issues are addressed now, please (re-)review.

Thank you!
